wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/E2EINotificationActionsHandler.swift

```diff
@@ -0,0 +1,198 @@
+
+    func getCertificate() async
+    func updateCertificate() async
+    func snoozeReminder() async
+
+}
+
+final class E2EINotificationActionsHandler: E2EINotificationActions {
+
+    // MARK: - Properties
+
+    private var enrollCertificateUseCase: EnrollE2EICertificateUseCaseProtocol
+    private var snoozeCertificateEnrollmentUseCase: SnoozeCertificateEnrollmentUseCaseProtocol
+    private var stopCertificateEnrollmentSnoozerUseCase: StopCertificateEnrollmentSnoozerUseCaseProtocol
+    private let gracePeriodRepository: GracePeriodRepository
+    private let targetVC: UIViewController
+
+    // MARK: - Life cycle
+
+    init(
+        enrollCertificateUseCase: EnrollE2EICertificateUseCaseProtocol,
+        snoozeCertificateEnrollmentUseCase: SnoozeCertificateEnrollmentUseCaseProtocol,
+        stopCertificateEnrollmentSnoozerUseCase: StopCertificateEnrollmentSnoozerUseCaseProtocol,
+        gracePeriodRepository: GracePeriodRepository,
+        targetVC: UIViewController) {
+            self.enrollCertificateUseCase = enrollCertificateUseCase
+            self.snoozeCertificateEnrollmentUseCase = snoozeCertificateEnrollmentUseCase
+            self.stopCertificateEnrollmentSnoozerUseCase = stopCertificateEnrollmentSnoozerUseCase
+            self.gracePeriodRepository = gracePeriodRepository
+            self.targetVC = targetVC
+        }
+
+    public func getCertificate() async {
+        let oauthUseCase = OAuthUseCase(rootViewController: targetVC)
+        do {
+            try await enrollCertificateUseCase.invoke(authenticate: oauthUseCase.invoke)
+            await confirmSuccessfulEnrollment()
+        } catch {
+            guard let endOfGracePeriod = gracePeriodRepository.fetchGracePeriodEndDate() else {
+                return
+            }
+            await showGetCertificateErrorAlert(canCancel: !endOfGracePeriod.isInThePast)
+        }
+    }
+
+    public func updateCertificate() async {
+        // TODO: [WPB-3324] update certificate
+    }
+
+    public func snoozeReminder() async {
+        guard let endOfGracePeriod = gracePeriodRepository.fetchGracePeriodEndDate(),
+              endOfGracePeriod.timeIntervalSinceNow > 0,
+              let formattedDuration = durationFormatter.string(from: endOfGracePeriod.timeIntervalSinceNow) else {
+            return
+        }
+
+        let alert = await UIAlertController.reminderGetCertificate(timeLeft: formattedDuration) {
+            Task {
+                await self.snoozeCertificateEnrollmentUseCase.invoke()
+            }
+        }
+        await targetVC.present(alert, animated: true)
+    }
+
+    // MARK: - Helpers
+
+    private func showGetCertificateErrorAlert(canCancel: Bool) async {
+        let oauthUseCase = OAuthUseCase(rootViewController: targetVC)
+        let alert = await UIAlertController.getCertificateFailed(canCancel: canCancel) {
+            Task {
+                try await self.enrollCertificateUseCase.invoke(authenticate: oauthUseCase.invoke)
```

Comment:
We should really be careful not missing issues like this:
Once there is a `try` in the closure without surrounding `do/catch` the `Task` instance has the throwing type `Task<Void, Error>`. But since no reference is kept, any error thrown will never be noticed.

Reply:
https://github.com/wireapp/wire-ios/pull/1051